server/channels/store/sqlstore/store.go

```diff
@@ -455,32 +471,93 @@ func (ss *SqlStore) GetSearchReplicaX() *sqlxDBWrapper {
 		return ss.GetReplicaX()
 	}
 
-	rrNum := atomic.AddInt64(&ss.srCounter, 1) % int64(len(ss.searchReplicaXs))
-	return ss.searchReplicaXs[rrNum]
+	cnt := 0
+	for {
+		rrNum := atomic.AddInt64(&ss.srCounter, 1) % int64(len(ss.searchReplicaXs))
+		if ss.searchReplicaXs[rrNum].Load().Online() {
+			return ss.searchReplicaXs[rrNum].Load()
+		}
+		cnt++
+		// If all replicas are down, then go with replica.
+		if cnt == len(ss.searchReplicaXs) {
+			break
+		}
+		// Check next
+		continue
+	}
+
+	return ss.GetReplicaX()
 }
 
 func (ss *SqlStore) GetReplicaX() *sqlxDBWrapper {
 	if len(ss.settings.DataSourceReplicas) == 0 || ss.lockedToMaster || !ss.hasLicense() {
 		return ss.GetMasterX()
 	}
 
-	rrNum := atomic.AddInt64(&ss.rrCounter, 1) % int64(len(ss.ReplicaXs))
-	return ss.ReplicaXs[rrNum]
-}
+	cnt := 0
+	for {
+		rrNum := atomic.AddInt64(&ss.rrCounter, 1) % int64(len(ss.ReplicaXs))
+		if ss.ReplicaXs[rrNum].Load().Online() {
+			return ss.ReplicaXs[rrNum].Load()
+		}
+		cnt++
+		// If all replicas are down, then go with master.
+		if cnt == len(ss.ReplicaXs) {
+			break
+		}
+		// Check next
+		continue
+	}
+	return ss.GetMasterX()
+}
+
+func (ss *SqlStore) monitorReplicas() {
+	t := time.NewTicker(time.Duration(*ss.settings.ReplicaMonitorIntervalSeconds) * time.Second)
+	defer func() {
+		t.Stop()
+		ss.wgMonitor.Done()
+	}()
+	for {
+		select {
+		case <-ss.quitMonitor:
+			return
+		case <-t.C:
+			setupReplica := func(r *atomic.Pointer[sqlxDBWrapper], dsn, name string) {
+				if r.Load().Online() {
+					return
+				}
+
+				handle, err := SetupConnection(name, dsn, ss.settings, 1)
```

Comment:
Would it make sense if we also make this attempt number configurable?

Reply:
IMO, I wouldn't want to. This anyways runs in a loop with the interval controlled by the `ReplicaMonitorIntervalSeconds`. If a customer wants fast recovery, they can set it to a low number. Or if they want slow recovery, they can increase it. That should be the setting to tune.

Reply:
Makes sense, thanks.